string mode
int8 dir
